samples/go/face-merge/main.go

```diff
@@ -0,0 +1,169 @@
+	"github.com/attic-labs/noms/go/walk"
+	flag "github.com/juju/gnuflag"
+)
+
+type facePoints struct {
+	x    types.Number
+	y    types.Number
+	w    types.Number
+	h    types.Number
+	cx   types.Number
+	cy   types.Number
+	name types.String
+}
+
+func main() {
+	mergeFaces()
+}
+
+func mergeFaces() {
+
+	var dbStr = flag.String("db", "", "input database spec")
+	var outDSStr = flag.String("out-ds", "", "output dataset to write to - if empty, defaults to input dataset")
+	verbose.RegisterVerboseFlags(flag.CommandLine)
+
+	flag.Usage = usage
+	flag.Parse(false)
+
+	if flag.NArg() == 0 {
+		flag.Usage()
+		return
+	}
+
+	cfg := config.NewResolver()
+	db, err := cfg.GetDatabase(*dbStr)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "Invalid input database '%s': %s\n", flag.Arg(0), err)
+		return
+	}
+	defer db.Close()
+
+	var outDS datas.Dataset
+	if !datas.IsValidDatasetName(*outDSStr) {
+		fmt.Fprintf(os.Stderr, "Invalid output dataset name: %s\n", *outDSStr)
+		return
+	}
+
+	outDS = db.GetDataset(*outDSStr)
+	inputs, err := spec.ReadAbsolutePaths(db, flag.Args()...)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "%s\n", err)
+		return
+	}
+
+	sizeType := types.MakeStructTypeFromFields("", types.FieldMap{
+		"width":  types.NumberType,
+		"height": types.NumberType,
+	})
+
+	dateType := types.MakeStructTypeFromFields("Date", types.FieldMap{
+		"nsSinceEpoch": types.NumberType,
+	})
+
+	photoType := types.MakeStructTypeFromFields("Photo", types.FieldMap{
+		"sizes":         types.MakeMapType(sizeType, types.StringType),
+		"title":         types.StringType,
+		"datePublished": dateType,
+		"dateUpdated":   dateType,
+	})
+
+	annotatedPhotoSet := types.NewGraphBuilder(db, types.SetKind, true)
+
+	for _, v := range inputs {
+		walk.WalkValues(v, db, func(cv types.Value) bool {
+			if types.IsSubtype(photoType, cv.Type()) {
+				s := cv.(types.Struct)
+				faces := getMergedFaces(s)
+				photo := filterExtraFields(s)
+				photo = photo.Set("faces", faces)
+				annotatedPhotoSet.SetInsert(nil, photo)
+			}
+			return false
+		})
+	}
+
+	outDS, err = db.Commit(outDS, annotatedPhotoSet.Build(), datas.CommitOptions{
+		Meta: types.NewStruct("", types.StructData{
+			"date": types.String(time.Now().Format(time.RFC3339)),
+		}),
+	})
+}
+
+func filterExtraFields(photo types.Struct) types.Struct {
+	filteredPhoto := types.NewStruct("Photo", types.StructData{})
+	photo.Type().Desc.(types.StructDesc).IterFields(func(fieldName string, t *types.Type) {
+		if fieldName != "facesCentered" && fieldName != "facesRect" {
+			filteredPhoto = filteredPhoto.Set(fieldName, photo.Get(fieldName))
+		}
+	})
+	return filteredPhoto
+}
+
+func getMergedFaces(photo types.Struct) types.Set {
+	facesRect, rectExists := photo.MaybeGet("facesRect")
+	facesCentered, centerExists := photo.MaybeGet("facesCentered")
+	faceSet := types.NewSet()
+
+	if rectExists && centerExists {
+		facesCentered.(types.Set).Iter(func(faceCentered types.Value) bool {
```

Comment:
Right here there is no guarantee that `facesCentered` is `types.Set`. In practice, this will work with our data, but it would be more self-documenting to actually do a type-check. The pattern I have been doing is like:

```
withFaces := types.NewStructWithFields("Photo", types.FieldData{
  "faces": ...
})
...
if (types.IsSubtype(photo.Type, withFaces) {
  ...
}
```